components/private_channel/rust/ffi/private-channel/src/lib.rs

```diff
@@ -0,0 +1,84 @@
+extern crate curve25519_dalek;
+extern crate sha2;
+
+use curve25519_dalek::scalar::Scalar;
+use elgamal_ristretto::ciphertext::Ciphertext;
+use elgamal_ristretto::private::SecretKey;
+use repsys_crypto::{
+    combine_pks, encrypt_input, generate_keys, partial_decryption_and_proof, randomize_and_prove,
+};
+
+use sha2::Sha512;
+use std::vec::Vec;
+
+use bincode;
+
+pub fn start_challenge(
+    input: Vec<String>,
+    server_pk_encoded: [u8; 32],
+) -> Result<(Vec<u8>, Vec<u8>, Vec<u8>, Vec<u8>), ()> {
```

Comment:
3+ items in a tuple starts to get confusing, especially if they're all the same type - a simple struct with public named fields would be much easier to understand and keep track of.

You can still use the existing semantics at the call site, but with better type safety, e.g. if you return something like this:

```rust
pub struct ChallengeOutput {
    pub pkey: Vec<u8>,
    pub skey: Vec<u8>,
    pub shared_pk: Vec<u8>,
    pub enc_hashes: Vec<u8>,
}
```

the callsite in the FFI can be written like this:

```rust
let ChallengeOutput { pkey, skey, shared_pk, enc_hashes } =
    match brave_private_channel::start_challenge(v_out, server_pk) {
        Ok(output) => output,
        Err(_) => return ResultChallenge::default(),
    };
```

Reply:
- [ ] `start_challenge` returns type instad of tuple